lib/archethic.ex

```diff
@@ -59,104 +60,49 @@ defmodule Archethic do
   @doc """
   Send a new transaction in the network to be mined. The current node will act as welcome node
   """
-  @spec send_new_transaction(Transaction.t(), Crypto.key(), nil | Contract.Context.t()) :: :ok
-  def send_new_transaction(
-        tx = %Transaction{},
-        welcome_node_key \\ Crypto.first_node_public_key(),
-        contract_context \\ nil
-      ) do
-    if P2P.authorized_and_available_node?() do
-      case NetworkChain.verify_synchronization(:node_shared_secrets) do
-        :ok ->
-          do_send_transaction(tx, welcome_node_key, contract_context)
-
-        :error ->
-          forward_transaction(tx, welcome_node_key, contract_context)
-
-        {:error, addresses} ->
-          SharedSecrets.genesis_address(:node_shared_secrets) |> SelfRepair.resync(addresses, [])
-
-          forward_transaction(tx, welcome_node_key, contract_context)
-      end
-    else
-      # node not authorized
-      forward_transaction(tx, welcome_node_key, contract_context)
-    end
-  end
+  @spec send_new_transaction(Transaction.t(), opts :: Keyword.t()) :: :ok
+  def send_new_transaction(tx = %Transaction{}, opts \\ []) do
+    welcome_node_key = Keyword.get(opts, :welcome_node_key, Crypto.first_node_public_key())
+    contract_context = Keyword.get(opts, :contract_context, nil)
+    forward? = Keyword.get(opts, :forward?, false)
 
-  @spec forward_transaction(
-          tx :: Transaction.t(),
-          welcome_node_key :: Crypto.key(),
-          contract_context :: Contract.Context.t()
-        ) :: :ok
-  defp forward_transaction(tx, welcome_node_key, contract_context) do
-    %Node{network_patch: welcome_node_patch} = P2P.get_node_info!(welcome_node_key)
+    if P2P.authorized_and_available_node?() and shared_secret_synced?() do
+      validation_nodes = Mining.get_validation_nodes(tx)
 
-    nodes =
-      P2P.authorized_and_available_nodes()
-      |> Enum.reject(&(&1.first_public_key == welcome_node_key))
-      |> Enum.sort_by(& &1.first_public_key)
-      |> P2P.nearest_nodes(welcome_node_patch)
-      |> Enum.filter(&P2P.node_connected?/1)
+      responses = do_send_transaction(tx, validation_nodes, welcome_node_key, contract_context)
 
-    this_node = Crypto.first_node_public_key()
+      maybe_start_resync(responses)
 
-    nodes =
-      if this_node != welcome_node_key do
-        #  if this node is not the welcome node then select
-        # next node from the this node position in nodes list
-        index = Enum.find_index(nodes, &(&1.first_public_key == this_node))
-        {_l, r} = Enum.split(nodes, index + 1)
-        r
-      else
-        nodes
+      if should_forward_transaction?(forward?, responses, length(validation_nodes)) do
```

Comment:
Here a suggestion to refactor a bit to simplify the forward should_forward_transaction

```suggestion
      forward? = case Keyword.get(opts, :forward?) do
        true ->
          should_forward_transaction?(responses, length(validation_nodes))
        _ ->
          false
      end
      
      maybe_start_resync(responses)

      if forward? do
        forward_transaction(tx, welcome_node_key, contract_context)
      end
```